File: src/lib/notion-task-fields.ts
import type { PageObjectResponse } from "@notionhq/client/build/src/api-endpoints.js";

/**
 * Task database property names (case-sensitive). Match your Notion DB after converting
 * from projects: if Notion renamed "Deadline" → keep `due` as the actual column name.
 */
export const TASK_PROPS = {
  title: "Name",
  due: "Due",
  status: "Status",
  project: "Project",
} as const;

export type TaskRowBrief = {
  id: string;
  name: string;
  status: string;
  due: string | null;
  projectPageId: string | null;
  url: string;
};

function titleFromPage(page: PageObjectResponse): string {
  for (const [, prop] of Object.entries(page.properties)) {
    if (
      prop &&
      typeof prop === "object" &&
      "type" in prop &&
      prop.type === "title" &&
      "title" in prop &&
      Array.isArray((prop as { title: { plain_text: string }[] }).title)
    ) {
      const title = (prop as { title: { plain_text: string }[] }).title;
      return (
        title.map((t: { plain_text: string }) => t.plain_text).join("") ||
        "Untitled"
      );
    }
  }
  return "Untitled";
}

export function mapTaskRow(page: PageObjectResponse): TaskRowBrief {
  const props = page.properties;
  const name = titleFromPage(page);
  let status = "";
  const st = props[TASK_PROPS.status];
  if (st?.type === "status") status = st.status?.name ?? "";
  else if (st?.type === "select") status = st.select?.name ?? "";
  let due: string | null = null;
  const d = props[TASK_PROPS.due];
  if (d?.type === "date") due = d.date?.start ?? null;
  let projectPageId: string | null = null;
  const rel = props[TASK_PROPS.project];
  if (rel?.type === "relation" && rel.relation[0])
    projectPageId = rel.relation[0].id;
  return { id: page.id, name, status, due, projectPageId, url: page.url };
}

File: src/lib/notion_task_fields.py
from typing import Any

TASK_PROPS = {
    "title": "Name",
    "due": "Due",
    "status": "Status",
    "project": "Project",
}


def title_from_page(page: dict[str, Any]) -> str:
    for prop in (page.get("properties") or {}).values():
        if isinstance(prop, dict) and prop.get("type") == "title":
            return "".join(t.get("plain_text", "") for t in prop.get("title") or []) or "Untitled"
    return "Untitled"


def map_task_row(page: dict[str, Any]) -> dict[str, Any]:
    props = page.get("properties") or {}
    status = ""
    st = props.get(TASK_PROPS["status"])
    if isinstance(st, dict):
        if st.get("type") == "status":
            status = (st.get("status") or {}).get("name") or ""
        elif st.get("type") == "select":
            status = (st.get("select") or {}).get("name") or ""
    due = None
    due_prop = props.get(TASK_PROPS["due"])
    if isinstance(due_prop, dict) and due_prop.get("type") == "date":
        due = (due_prop.get("date") or {}).get("start")
    project_page_id = None
    rel = props.get(TASK_PROPS["project"])
    if isinstance(rel, dict) and rel.get("type") == "relation" and rel.get("relation"):
        project_page_id = rel["relation"][0].get("id")
    return {
        "id": page.get("id"),
        "name": title_from_page(page),
        "status": status,
        "due": due,
        "projectPageId": project_page_id,
        "url": page.get("url"),
    }
